File: memories/2024/exercise/scala/day13/src/main/scala/santamarket/model/ChristmasElf.scala
package santamarket.model

import scala.collection.mutable

class ChristmasElf(private val catalog: SantamarketCatalog) {
  private val offers = mutable.Map[Product, Offer]()

  def addSpecialOffer(offerType: SpecialOfferType, product: Product, argument: Double): Unit = {
    offers(product) = Offer(offerType, product, argument)
  }

  def checksOutArticlesFrom(sleigh: ShoppingSleigh): Receipt = {
    val receipt = new Receipt()

    for (item <- sleigh.getItems) {
      val p = item.product
      val quantity = item.quantity
      val unitPrice = catalog.getUnitPrice(p)
      receipt.addProduct(quantity * unitPrice)
    }

    sleigh.handleOffers(receipt, offers.toMap, catalog)

    receipt
  }
}

File: memories/2024/exercise/C#/day13/SantaMarket/Model/ChristmasElf.cs
using System.Collections.Generic;

namespace SantaMarket.Model
{
    public class ChristmasElf
    {
        private readonly ISantamarketCatalog catalog;
        private readonly Dictionary<Product, Offer> offers = new Dictionary<Product, Offer>();

        public ChristmasElf(ISantamarketCatalog catalog)
        {
            this.catalog = catalog;
        }

        public void AddSpecialOffer(SpecialOfferType offerType, Product product, double argument)
        {
            offers[product] = new Offer(offerType, product, argument);
        }

        public Receipt ChecksOutArticlesFrom(ShoppingSleigh sleigh)
        {
            var receipt = new Receipt();

            foreach (var item in sleigh.GetItems())
            {
                var p = item.Product;
                var quantity = item.Quantity;
                var unitPrice = catalog.GetUnitPrice(p);
                receipt.AddProduct(quantity * unitPrice);
            }

            sleigh.HandleOffers(receipt, offers, catalog);

            return receipt;
        }
    }
}
